src/cpp/fastrtps_deprecated/publisher/PublisherHistory.cpp

```diff
@@ -161,7 +171,11 @@ bool PublisherHistory::add_pub_change(
                 }
                 else
                 {
-                    logWarning(RTPS_HISTORY, "Change not added due to maximum number of samples per instance");
+                    CacheChange_t* change_to_remove = vit->second.cache_changes.front();
+                    if (mp_writer->wait_for_acknowledgement(change_to_remove, max_blocking_time, lock))
```

Comment:
It would be more secure if this function receives a SequenceNumber_t instead of a CacheChange_t pointer.